notebooks/autodiff_cookbook.ipynb

```diff
@@ -0,0 +1,1062 @@
+        {
+          "output_type": "stream",
+          "text": [
+            "Non-vmapped Jacobian-Matrix product\n",
+            "10 loops, best of 3: 67.7 ms per loop\n",
+            "\n",
+            "Vmapped Jacobian-Matrix product\n",
+            "10 loops, best of 3: 5.5 ms per loop\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "metadata": {
+        "id": "MXFEFBDz6vvL",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### The implementation of `jacfwd` and `jacrev`\n",
+        "\n",
+        "Cool, cool, we got this. *Autograd couldn't do this.*"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "HBEzsDH1U5_4",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import jacrev as builtin_jacrev\n",
+        "\n",
+        "def our_jacrev(f):\n",
+        "  def jacfun(x):\n",
+        "    y, vjp_fun = vjp(f, x)\n",
+        "    # Use vmap to do a matrix-Jacobian product.\n",
+        "    # Here, the matrix is the Euclidean basis, so we get all\n",
+        "    # entries in the Jacobian at once. \n",
+        "    J, = vmap(vjp_fun, in_axes=0)(np.eye(len(y)))\n",
+        "    return J\n",
+        "  return jacfun\n",
+        "\n",
+        "assert np.allclose(builtin_jacrev(f)(W), our_jacrev(f)(W)), 'Incorrect reverse-mode Jacobian results!'"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "Qd9gVZ5t6vvP",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import jacfwd as builtin_jacfwd\n",
+        "\n",
+        "def our_jacfwd(f):\n",
+        "  def jacfun(x):\n",
+        "    _jvp = lambda s: jvp(f, (x,), (s,))[1]\n",
+        "    Jt =vmap(_jvp, in_axes=1)(np.eye(len(x)))\n",
+        "    return np.transpose(Jt)\n",
+        "  return jacfun\n",
+        "\n",
+        "assert np.allclose(builtin_jacfwd(f)(W), our_jacfwd(f)(W)), 'Incorrect forward-mode Jacobian results!'"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "YDOD1NUT6vvT",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### The implementation of `hessian`"
+      ]
+    },
+    {
+      "metadata": {
+        "colab_type": "code",
+        "id": "Tinc6qKK-hwE",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import hessian as builtin_hessian\n",
+        "\n",
+        "def our_hessian(f):\n",
+        "  # Forward-over-reverse calculation of the Hessian.\n",
+        "  # We could also do jacfwd(jacfwd(W)), jacrev(jacrev(W)) or jacrev(jacfwd(W)).\n",
+        "  # They're all equally valid. Try them out and time them!\n",
+        "  return jacfwd(jacrev(f))\n",
```

Comment:
Our usual formulation also sticks a `jit()` around this, you may wish to do the same or add text observing that you can so do.

Reply:
Added a version of this.